### ISSUE TITLE
Happy Eyeballs crashes with Eager Task Factory

### ISSUE THREAD

**Issue by mhils** [python/cpython#124309]
# Bug report

### Bug description:

asyncio's `staggered_race` crashes when an eager task factory is being used:

```python
import asyncio

async def main():
    asyncio.get_running_loop().set_task_factory(asyncio.eager_task_factory)
    await asyncio.open_connection("example.com", 80, happy_eyeballs_delay=0.25)

asyncio.run(main())
```
```
Traceback (most recent call last):
  [...]
  File "/tmp/repro.py", line 5, in main
    await asyncio.open_connection("example.com", 80, happy_eyeballs_delay=0.25)
  File "/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/streams.py", line 48, in open_connection
    transport, _ = await loop.create_connection(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/base_events.py", line 1109, in create_connection
    sock, _, _ = await staggered.staggered_race(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/staggered.py", line 144, in staggered_race
    raise d.exception()
  File "/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/staggered.py", line 101, in run_one_coro
    assert len(running_tasks) == this_index + 2
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

See also: https://github.com/aio-libs/aiohttp/issues/8599

### CPython versions tested on:

3.12

### Operating systems tested on:

macOS

### Linked PRs
* gh-124311
* gh-124390
* gh-124573
* gh-124574
* gh-124810
* gh-124817
* gh-124847
* gh-125339
* gh-125340

**Comment by rruuaanng** [comment]
I'd love to help you, but what I can't understand is why it creates an extra task that triggers the assertion.
```
next_task = loop.create_task(run_one_coro(this_failed))
running_tasks.append(next_task)
print('\n', 'HELLO', len(running_tasks), this_index)
assert len(running_tasks) == this_index + 2
# Prepare place to put this coroutine's exceptions if not won
exceptions.append(None)
assert len(exceptions) == this_index + 1
```
```
 HELLO 1 1
 HELLO 3 2
```
Maybe you can start from here, sorry I'm not familiar with this.

**Comment by rruuaanng** [comment]
Hmmm, I submitted a PR, but it doesn't seem to solve the problem any better, I'm looking for.

**Comment by kumaraditya303** [comment]
Probably removing that assert is sufficient.

**Comment by rruuaanng** [comment]
> Probably removing that assert is sufficient.

Removing that assertion won't fix the issue; instead, it could cause new errors elsewhere. the real problem isn't with the assertion anyway.

**Comment by ZeroIntensity** [comment]
This issue is unrelated to Happy Eyeballs, but instead that `staggered_race` doesn't support `eager_task_factory`. Seperate reproducer:

```py
import asyncio
import random

async def main():
    asyncio.get_running_loop().set_task_factory(asyncio.eager_task_factory)
    await asyncio.staggered.staggered_race([asyncio.sleep(0) for _ in range(5)], 0.25)

asyncio.run(main())
```

Using -O (to disable assertions) is a workaround for all versions, but maybe it's worth revisiting `staggered_race` a little bit to make sure it's working correctly with eager task factory. I'm working on that.

**Comment by ZeroIntensity** [comment]
Ah, using `delay=None` in my reproducer causes a deadlock. Removing the assertion isn't 100% effective.

**Comment by ZeroIntensity** [comment]
I've created #124390 as a fix. The existing implementation was so broken for eager task factories (due to the reliance of tasks not starting before you `await` them), so I had to pretty much rewrite it.

Could you confirm that it solves your problem, @mhils?

**Comment by rruuaanng** [comment]
> I've created #124390 as a fix. The existing implementation was so broken for eager task factories (due to the reliance of tasks not starting before you `await` them), so I had to pretty much rewrite it.
> 
> Could you confirm that it solves your problem, @mhils?

That's right, I was confused when I tried to fix it, I kept thinking that maybe it was supporting `happy_eyeballs`.

**Comment by ZeroIntensity** [comment]
Sorry, we had to revert the fix due to the complexity of it (and also because it was causing some other problems). We'll (hopefully) have the fix for this issue in the next-next 3.12 patch (not tomorrow's, the one after, so 3.12.8).

**Comment by graingert** [comment]
I think this can be closed, it's backported to 3.12 and 3.13
